iio: dac: ad5758: Use slew time to determine the slew rate step and c…

stefpopa opened the pull request:
The slew time can be expressed as a function of the step size and the
update clock. Therefore, a slew time entry replaces the slew rate step and
slew rate clock entries which are now determined at start up with a best
match algorithm.

If the slew time entry is not specified, then the slew rate control is not
enabled.

Signed-off-by: Stefan Popa <stefan.popa@analog.com>

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -351,22 +367,85 @@ static int ad5758_set_dc_dc_ilim(struct ad5758_state *st, unsigned int ilim)
 	unsigned long int mask;
-	int ret;
+	int *index, i, sr_clk_index, sr_step_index, ret;
+	long long diff_new, diff_old;
```
should we use `int64_t`  and `uint64_t` instead ?
i'm always paranoid when i see `long long` and `unsigned long long`

stefpopa replied:
I can use u64 and s64, but it would be the only place in the driver where I do this. For consistency purposes, I thought not to mix them up.

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -351,22 +367,85 @@ static int ad5758_set_dc_dc_ilim(struct ad5758_state *st, unsigned int ilim)
+		 * Step Size = Full Scale Out / (Update Clk Freq * Slew Time)
+		 */
+		sr_step_tmp = AD5758_FULL_SCALE_MICRO;
+		do_div(sr_step_tmp, ad5758_slew_rate_clk[i]);
```
does `do_div` handle 64 bit values ?
would `do_div64` work better here ?

stefpopa replied:
do_div does handle 64 bit values; I haven't seen any uses of do_div64

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -351,22 +367,85 @@ static int ad5758_set_dc_dc_ilim(struct ad5758_state *st, unsigned int ilim)
+
+	sr_clk_index = 15 - (index - ad5758_slew_rate_clk);
+
+	index = (int *) bsearch(&sr_step,
```
an idea would be to wrap this logic into it's own search function that finds `sr_step_index` & `sr_clk_index`

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -351,22 +367,85 @@ static int ad5758_set_dc_dc_ilim(struct ad5758_state *st, unsigned int ilim)
-	mask = (AD5758_DAC_CONFIG_SR_EN_MSK |
-	       AD5758_DAC_CONFIG_SR_CLOCK_MSK |
-	       AD5758_DAC_CONFIG_SR_STEP_MSK);
+	diff_old = sizeof(long long);
```
is there something like INT_MAX or INT64_MAX or similar to use here ?

stefpopa replied:
I don't know :)

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -351,22 +367,85 @@ static int ad5758_set_dc_dc_ilim(struct ad5758_state *st, unsigned int ilim)
+		 * is different from the desired slew time and store the diff
+		 * for the next iteration
+		 */
+		diff_new = st->slew_time - calc_slew_time;
```
maybe do `diff_new = abs(st->slew_time - calc_slew_time);` here

stefpopa replied:
I agree

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -351,22 +367,85 @@ static int ad5758_set_dc_dc_ilim(struct ad5758_state *st, unsigned int ilim)
+		 * After a raw value for step size was determined, find the
+		 * closest valid match
+		 */
+		sr_step_index = ad5758_get_array_index(ad5758_slew_rate_step,
```
this looks like it needs an exact match for slew rate step;
if not found, you get `-EINVAL`;
is this ok ? or would it be better to look for a closest match here ?

stefpopa replied:
You are right! It's not ok, I need a closest match function instead.

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -351,22 +367,85 @@ static int ad5758_set_dc_dc_ilim(struct ad5758_state *st, unsigned int ilim)
+	if (index == NULL)
+		return -ENOENT;
+
+	sr_clk_index = 15 - (index - ad5758_slew_rate_clk);
```
not sure if adding `15` is the best idea; maybe `ARRAY_SIZE(ad5758_slew_rate_clk) - 1` or similar

stefpopa replied:
I agree

stefpopa commented:
changelog v1 -> v2:
* Added a new commit which replaces the use of ad5758_get_array_index() with bsearch()
* used a new ad5758_find_closest_match() function
* a new ad5758_slew_rate_set() function takes as input the desired slew rate clock and step, searches for the index in the array and sets the registers
* used UINT_MAX to initialize diff_old variable
* used s64 and u64 for better readability inside the ad5758_slew_rate_config() function
* used ARRAY_SIZE(ad5758_slew_rate_clk) - 1 instead of the "magic number" 15
* diff_new = abs(st->slew_time - calc_slew_time);

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -371,6 +415,57 @@ static int ad5758_slew_rate_config(struct ad5758_state *st,
+	s64 diff_new, diff_old;
+	u64 sr_step_tmp, calc_slew_time;
+
+	diff_old = UINT_MAX;
```
maybe use `S64_MAX` here;

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -239,16 +250,22 @@ static int ad5758_spi_write_mask(struct ad5758_state *st,
-	for (i = 0; i < size; i++) {
-		if (val == array[i])
-			return i;
+	for (i = 1; i < size; i++) {
```
typically (i-1) this is used when you compare 2 indices
maybe, start from i = 0; and return  (size - 1) if nothing found;
that way it's easier to understand;

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -377,6 +415,57 @@ static int ad5758_slew_rate_config(struct ad5758_state *st,
+		if (diff_new < diff_old) {
+			diff_old = diff_new;
+			sr_clk = ad5758_slew_rate_clk[i];
+			sr_step = ad5758_slew_rate_step[sr_step_index];
```
since you have the indexes here for sr_clk & sr_step_index, do you need to do the bsearch calls in `ad5758_slew_rate_set()` ?
maybe, you could just pass the indexes ; it would save some lines of code;

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -637,19 +726,20 @@ static int ad5758_crc_disable(struct ad5758_state *st)
 				 "dc-dc-ilim out of range using default");
 		else
-			st->dc_dc_ilim = index;
+			st->dc_dc_ilim = index - ad5758_dc_dc_ilimt;
```
this looks incorrect in one way or another;
`st->dc_dc_ilim` is initialized to 150, and this stores the index/position of the value ;

stefpopa commented:
Changelog v3 -> v4:
* squashed the two commits

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -351,22 +368,22 @@ static int ad5758_set_dc_dc_ilim(struct ad5758_state *st, unsigned int ilim)
-	mask = (AD5758_DAC_CONFIG_SR_EN_MSK |
-	       AD5758_DAC_CONFIG_SR_CLOCK_MSK |
-	       AD5758_DAC_CONFIG_SR_STEP_MSK);
+	sr_clk_index = (ARRAY_SIZE(ad5758_sr_clk) - 1) - sr_clk_index;
```
do you need to still do this adjustment for the index ?

stefpopa replied:
actually, if bsearch() is not used anymore to find the index, then ad5758_sr_clk doesn't need to be in increasing order anymore, therefore can be defined as: 
```
static const int ad5758_sr_clk[16] = {
	240000, 200000, 150000, 128000, 64000, 32000, 16000, 8000, 4000,
	2000, 1000, 512, 256, 128, 64, 16
};
```

stefpopa commented:
changelog v4 -> v5:
* remove the adjustment for sr_clk_index 
* remove the unused struct ad5758_slew_rate

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -172,12 +177,12 @@ static const struct ad5758_range ad5758_min_max_table[] = {
 };
 
-static const int ad5758_slew_rate_clk[16] = {
+static const int ad5758_sr_clk[16] = {
```
nitpick/curios: any reason to change these names ?

stefpopa replied:
I thought the name is a bit too long and also going through these arrays I am using variables called sr_step_index and sr_clk_index

commodo replied:
ok; let's leave these shorter than

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -351,22 +362,20 @@ static int ad5758_set_dc_dc_ilim(struct ad5758_state *st, unsigned int ilim)
-		AD5758_DAC_CONFIG_SR_STEP_MODE(sr_step));
+	       AD5758_DAC_CONFIG_SR_STEP_MSK;
+	mode = AD5758_DAC_CONFIG_SR_EN_MODE(1) |
+		AD5758_DAC_CONFIG_SR_STEP_MODE(sr_step_index) |
```
do these 2 lines need to be re-indented ?

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -377,6 +386,57 @@ static int ad5758_slew_rate_config(struct ad5758_state *st,
+	unsigned int sr_clk_index, sr_step_index;
+	int i, sr_step_index_tmp;
+	s64 diff_new, diff_old;
+	u64 sr_step_tmp, calc_slew_time;
```
just styling comments:
* `sr_step_tmp` can be named just `sr_step` ; it doesn't look like the `_tmp` suffix is needed
* maybe compress `sr_clk_index` -> `sr_clk_idx` & `sr_step_index` -> `sr_step_idx` ; they feel a bit too long ; though, this is a preference, so feel free to disregard
* `sr_step_index_tmp` also feels a bit too long ; maybe just use `res` as a variable name ; it should be easy to understand it's purpose, since it's use is fairly compact

commodo commented:
otherwise, this looks good from my side;

stefpopa commented:
changelog v5 -> v6:
* renamed a few variables: `sr_step_tmp `-> `sr_step `; `sr_clk_index `-> `sr_clk_idx `& `sr_step_index `-> `sr_step_idx `;  `sr_step_index_tmp `-> `res `
* fixed indentation
